sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -100,6 +101,15 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     withOrigin(ctx)(StructType(visitColTypeList(ctx.colTypeList)))
   }
 
+  override def visitSingleInterval(ctx: SingleIntervalContext): CalendarInterval = {
+    withOrigin(ctx) {
+      val intervals = ctx.intervalField.asScala.map(visitIntervalField)
+      validate(intervals.nonEmpty,
+        "at least one time unit should be given for interval literal", ctx)
+      intervals.reduce(_.add(_))
```

Comment:
You create an instance of `CalendarInterval` per each interval units, and always have to summarize months and microseconds  (and maybe days in the future), and create new instance per each `add()`?

Let's benchmark this and see how it is fast.

Reply:
This is how it was done in the parser. It's indeed different from `CalendarInterval.fromCaseInsensitiveString`. Let me see how to follow `CalendarInterval` and make the parser more efficient.